src/java.base/share/classes/jdk/internal/math/ToDecimal.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (c) 2003, 2024, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2024, Alibaba Group Holding Limited. All Rights Reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package jdk.internal.math;
+
+import jdk.internal.access.JavaLangAccess;
+import jdk.internal.access.SharedSecrets;
+
+import static java.lang.Math.multiplyHigh;
+
+abstract sealed class ToDecimal permits DoubleToDecimal, FloatToDecimal {
+    private static final JavaLangAccess JLA = SharedSecrets.getJavaLangAccess();
+
+    /* Used for left-to-tight digit extraction */
+    static final int MASK_28 = (1 << 28) - 1;
+
+    static final int NON_SPECIAL = 0 << 8;
+    static final int PLUS_ZERO   = 1 << 8;
+    static final int MINUS_ZERO  = 2 << 8;
+    static final int PLUS_INF    = 3 << 8;
+    static final int MINUS_INF   = 4 << 8;
+    static final int NAN         = 5 << 8;
+
+    static final byte LATIN1 = 0;
```

Comment:
I think this somewhat unnecessarily copies names and internal implementation details from `String` (where the `coder` value is used both as a pseudo-boolean and as a shift operand, which is not applicable here). In this context we could use something like `private final boolean latin1;` (all uses of `coder` are just `if (coder == LATIN1)` so it would be simplified to `if (latin1)`)

Reply:
Yeah, I'm just musing about the perils of leaking/replicating those implementation details, but as you say this is internal code only used by `String` and `StringBuilder` to it's loosely part of that complex. Not requesting any changes here for now..

Reply:
I looked at it again and I think @cl4es 's suggestion is correct. Can I submit the change? @rgiulietti

Reply:
Of course, nothing has been approved as of now.

Since you are preparing a commit anyway, may I ask you to revert back the changes [here](https://github.com/openjdk/jdk/pull/19730#discussion_r1652978056)? Thanks.

Reply:
I have completed all the suggested changes. Please continue to review.